### PR TITLE
Adds LBFGS.

### PR DESCRIPTION
**Pull request by elibol**

### CODE REVIEW

File: nums/models/glms.py [hunk 1]
```diff
@@ -350,10 +382,6 @@ def hessian(self, X: BlockArray, y: BlockArray, mu: BlockArray = None):
 PoissonRegressor = PoissonRegression
 
 
-def line_search():
```
Comment (vinamrabenara) [inline nums/models/glms.py:353]: is line search now part of nums/models/lbfgs.py (BackTrackingLineSearch)?

Comment (elibol) [reply nums/models/glms.py:353]: Yes.

File: nums/models/glms.py [hunk 2]
```diff
@@ -462,8 +490,21 @@ def irls(
...
+def lbfgs(
+    app: ArrayApplication,
+    model: GLM,
+    beta,
+    X: BlockArray,
+    y: BlockArray,
+    tol: BlockArray,
+    max_iter: int,
+):
+    # TODO (hme): Enable a way to provide memory length and line search parameters.
```
Comment (vinamrabenara) [inline nums/models/glms.py:503]: Where exactly do we want to provide line search params. What kind of params would these be. would they be kind of line search, like backtracking, etc and max iters and other line search params like alpha etc..

File: tests/models/test_lbfgs.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+import numpy as np
+from nums.models.glms import LogisticRegression
+from nums.core.array.application import ArrayApplication
+
+
+def sample_set(app: ArrayApplication):
+    shape = (500, 10)
+    block_shape = (100, 10)
+    rs = app.random_state(1337)
```
Comment (vinamrabenara) [inline tests/models/test_lbfgs.py:25]: is 1337 a seed

Comment (elibol) [reply tests/models/test_lbfgs.py:25]: Yes.

File: nums/models/glms.py (hunk 2 above)
Comment (elibol) [reply nums/models/glms.py:503]: Yes, I think the best solution here is to allow a way to provide those params via the GLM class constructor `__init__`. For backtracking, you can see a number of parameters hardcoded in the LBFGS class, including initial alpha value, how much it is reduced per iteration, and the slope param.

**Pull request merged**